Create: Tracing docs (#1516)
getsentry/sentry-docs: src/collections/_documentation/apm/apm-glossary.md

---
title: APM Glossary
sidebar_order: 5
---

## Clock Skew

> If you are collecting transactions from more than a single machine, you will likely encounter **clock skew**. When collecting and comparing transaction events that are gathered from different devices, you may observe timestamps that do not align or events in a trace that don't share a time history. Sentry does not attempt to correct the timestamps in your events. The timestamps displayed in Sentry for transactions and each span retain the original values reported by your application/hosting environment.

Reviewer: ```suggestion
If you are collecting transactions from multiple machines, you will likely encounter **clock skew**. When collecting and comparing transaction events that are gathered from different devices, you may observe timestamps that do not align or events in a trace that don't share a time history. Sentry does not attempt to correct the timestamps in your events. The timestamps displayed in Sentry for transactions and each span retain the original values reported by your application/hosting environment.
```

While you can reduce clock skew by utilizing Network Time Protocol (NTP) or your cloud provider's clock synchronization services, you may still notice small drifts in your data, as synchronizing clocks on small intervals is challenging.

## Performance Metrics

> As part of APM and transaction data, we’re exposing a few new aggregate metrics to help customers better understand the performance of their applications.

Reviewer: ```suggestion
Some aggregate functions can be applied onto transaction events to help you better understand the performance characteristics of your applications.
```

Need to change the language of this to be more timeless.

### Duration Percentiles

> The average, 75%, 95%, and 99% duration, help guide customers in identifying transactions that are slower than the user’s target SLAs.

Reviewer: ```suggestion
You can aggregate transaction durations using the following aggregate functions:

- average
- 75%, 95%, and 99% duration percentiles 

One usecase for using these statistics is to help you identify transactions that are slower than your organization's target SLAs.
```

> Our APM data is sampled, so the percentiles we present only represent the data we have. Because of sampling, data ranges, filters, or a low traffic transaction, we will often have a case where data is directionally correct, but not accurate. For example, the average of one number is that number, but that’s not usually what people expect to see.

Reviewer: ```suggestion
Transaction events are sampled, so the percentiles we present within Sentry only represents the data we have. Because of sampling, data ranges, filters, or low volume of transactions, we will often have a case where data is directionally correct, but not accurate. For example, the average of one number is that number, but that’s not usually what you may expect to see.
```

This inability to be usefully accurate will happen more often in some columns than others. We can calculate an average with less data than a 95th percentile, but it’ll also vary by row. An index page will always get more traffic than `/settings/country/belgium/tax`.

### Percentage Query

This column shows the percentage of the entire query that this row represents, so you filter your results with the query and then see how much of the result set a particular transaction represents.

### Requests Per Minute (RPM)

Requests Per Minute is a way to measure throughput. It is the average of all requests bucketed by the minute for the current time window and query string.

## Span

> The **span** is the primary building block of a distributed trace, representing an individual unit of work done in a distributed system. It can be the processing of an HTTP request by a server, the querying of a database, or an HTTP call to an external service.

Reviewer: ```suggestion
The **span** is the primary building block of a distributed trace, representing an individual unit of work done in a distributed system. It may represent the processing of an HTTP request by a server, the querying of a database, or an HTTP call to an external service.
```

> An example of a span that describes a call to an external service:

Reviewer: @MimiDumpling this payload example may be better moved to the end of the "Span" section; after the paragraphs elaborating on the span concept.

Spans can have descendant spans (or child spans). In our earlier example, the three subroutines can be spans within a larger encompassing span (the database call).

Any instrumented span is part of a trace (identified by its trace id, `trace_id`), and each span has its own marker called the `span_id`, a fixed length of alphanumeric characters.

### Hierarchy

Each span can be connected to other spans using a parent-child hierarchy. 

For example, a span called `updateJiraTask` would have child spans like `Check permissions`, `Update Task`, `sql.update`, `send.webhooks`.

Each sub-task span would have an id in the attribute `parent_span_id` id of the `updateJiraTask` span.

### Orphan Spans

Orphan spans refer to spans that exist or are visible, but their parent spans are not visible or are not within Sentry's database. This could imply a bug in the Sentry application. If you're confident about this, please contact support: [support@sentry.io](mailto:support@sentry.io)

> Another reason could include that instrumented spans may not make it back to Sentry for processing. As an example, let us refer back to our earlier example of the web application consisting of:

Reviewer: ```suggestion
Another reason could include that instrumented spans may not make it back to Sentry for processing. As an example, let us consider the web application consisting of:
```

@MimiDumpling updating this since this is now in the glossary.

- Frontend (Single-Page Application)
- Backend
- Background Queue
- Notification Job

> Instrumented spans in each component are sent back to Sentry for processing. However, a server in the backend component could suffer from network connectivity loss, and the spans may not have made it back to Sentry. This would indicate a "hole" within this specific trace. The collected spans within the Search Service would be descendants of a span (for example, a search service call) within the Backend component, and thus the top-level span of the collective Search Service spans is referred to as an orphan span.

Reviewer: ```suggestion
Instrumented spans in each component are sent back to Sentry for processing. However, a server in the backend component could suffer from network connectivity loss, and the spans may not have made it back to Sentry. This would indicate a "hole" within this specific trace. The collected spans within the Background Queue would be descendants of a span (for example, a queue service call) within the Backend component, and thus the top-level span of the collective Background Queue spans is referred to as an orphan span.
```

Traces that start with the same set of API endpoint calls can potentially differ in terms of what code paths were executed. For example, an initial trace may result in a cache miss at the Cache Service, and a subsequent trace may include more information instrumented at the Cache Service. Although these two traces are similar, they are not identical.

### Propagation

Processing of the input data in the distributed environment means that an arbitrary number of the systems/microservices can be involved.

To connect all the spans that were created during the processing of the input data, we need to be able to propagate the basic identifier. By forwarding `trace_id` and `parent_span_id` as data travels through various applications in a microservice ecosystem, we can reconstruct data flows later.

Trace metadata is propagated through http headers that are included in the http request to the external system.

## Transaction

> There already was a concept of transaction in the Sentry SDKs before we started with APM.

Reviewer: ```suggestion
```

I think we can omit this line altogether.

Reviewer: @HazAT @untitaker was this something tracked by the SDKs itself prior to tracing?

Transaction is an event attribute that provides initial context to the error report. This makes it possible for the error to be linked to a specific part of the system. An example of the transaction name can be “/users/” which is the name for the endpoint that received the request that failed during the processing.

> With APM, the transaction is a parent span to all spans created during the processing of the request in the single server.

Reviewer: ```suggestion
With tracing, the transaction is a parent span to all spans created during the processing of the request in the single server.
```

---

> title: APM Glossary

Reviewer 2: Suggest rename to Performance Glossary

If you are collecting transactions multiple machines, you will likely encounter **clock skew**. When collecting and comparing transaction events that are gathered from different devices, you may observe timestamps that do not align or events in a trace that don't share a time history. Sentry does not attempt to correct the timestamps in your events. The timestamps displayed in Sentry for transactions and each span retain the original values reported by your application/hosting environment.

While you can reduce clock skew by utilizing Network Time Protocol (NTP) or your cloud provider's clock synchronization services, you may still notice small drifts in your data, as synchronizing clocks on small intervals is challenging.

## Performance Metrics

Some aggregate functions can be applied to transaction events to help you better understand the performance characteristics of your applications.

### Duration Percentiles

You can aggregate transaction durations using the following aggregate functions:
- average
- 75%, 95%, and 99% duration percentiles

One use-case for using these statistics is to help you identify transactions that are slower than your organization's target SLAs.

> Transaction events are sampled, so the percentiles we present within Sentry only represents the data we have. Because of sampling, data ranges, filters, or low volume of transactions, we will often have a case where data is directionally correct, but not accurate. For example, the average of one number is that number, but that’s not usually what you may expect to see.

Reviewer 2: Suggest "Transaction events are sampled. As a result, the percentiles presented within Sentry represent only the data received."

This inability to be usefully accurate will happen more often in some columns than others. We can calculate an average with less data than a 95th percentile, but it’ll also vary by row. An index page will always get more traffic than `/settings/country/belgium/tax`.

### Percentage Query

> This column shows the percentage of the entire query that this row represents, so you filter your results with the query and then see how much of the result set a particular transaction represents.

Author: ^ @dashed @bowencai8 ?

> This column shows the percentage of the entire query that this row represents, so you filter your results with the query and then see how much of the result set a particular transaction represents.

Author: ^ @dashed I think this is part of the Performance Metric Definitions?

> This column shows the percentage of the entire query that this row represents, so you filter your results with the query and then see how much of the result set a particular transaction represents.

Reviewer: @MimiDumpling 

It doesn't look like this section is being referenced anywhere. We can either omit or hide this content.

> This column shows the percentage of the entire query that this row represents, so you filter your results with the query and then see how much of the result set a particular transaction represents.

Author: Sounds good!

When you initialize and load a Sentry SDK that supports tracing, it will automatically instrument parts of your running application, such as function calls and network operations. Instrumenting a database call could entail these three steps or subroutines:

1. Calling a function to compose a database query
2. Sending a query to a database
3. Receiving and transforming results

Each of these subroutines takes some time to perform (no matter how fast they could run). Sentry records these subroutines as a unit of work in the form of **spans**. Sentry can also attach useful information to spans, such as tags, additional contextual data, and a status to indicate if the subroutine failed or not.

Spans can have descendant spans (or child spans). In our earlier example, the three subroutines can be spans within a larger encompassing span (the database call).

Any instrumented span is part of a trace (identified by its trace id, `trace_id`), and each span has its own marker called the `span_id`, a fixed length of alphanumeric characters.

> An example of a span that describes a call to an external service:

Author: @untitaker do we have a better code example using `createJiraTask`? If so, will you provide the code example?

Reviewer 3: I have no idea where createJiraTask comes from. @markstory or @dashed may know

Reviewer: I think @markstory came up with this. 

Would it be better to come up with another example, or recycle an example to keep this section consistent with the rest of the tracing docs?

Author: I'd try to keep it consistent with the other examples. If it's better communication to create a unique example, I'm okay with that option for those reasons.

Reviewer: @MimiDumpling 

Use this code sample:

```json

// updateJiraTask
{
    "trace_id": "a8233eb845154a2cae4712b28375d08d",
    "parent_span_id": "ba99c37bfb30f860",
    "span_id": "92ce7ea47d5b8a45",
    "start_timestamp": 1583868931.940406,
    "same_process_as_parent": true,
    "description": "updateJiraTask",
    "tags": { "status": "ok" },
    "timestamp": 1583868932.060615,
    "op": "jira",
    "data": {}
}

// sql.update - a child span of updateJiraTask
{
    "trace_id": "a8233eb845154a2cae4712b28375d08d",
    "parent_span_id": "92ce7ea47d5b8a45",
    "span_id": "a88e33dab5b745b9",
    "start_timestamp": 1583868931.953386,
    "same_process_as_parent": true,
    "description": "sql.update - SELECT * FROM jira_table",
    "tags": { "status": "ok" },
    "timestamp": 1583868932.03692,
    "op": "db",
    "data": {}
}
```

```json
    {
      "trace_id": "a0fa8803753e40fd8124b21eeb2986b5",
      "parent_span_id": "9c2a6db8c79068a2",
      "span_id": "8c931f4740435fb8",
      "start_timestamp": 1563890702.134,
      "same_process_as_parent": true,
      "description": "http://httpbin.org/base64/aGVsbG8gd29ybGQK GET",
      "tags": { "http.status_code": 200, "status": "ok" },

>       "timestamp": "1563890702.934",

Author: @untitaker should I change this timestamp?

Author: If so, what should I change it to?

Reviewer 3: Yeah please just drop the quotes around the number like for `start_timestamp`

- Frontend (Single-Page Application)
- Backend
- Background Queue
- Notification Job

Instrumented spans in each component are sent back to Sentry for processing. However, a server in the backend component could suffer from network connectivity loss, and the spans may not have made it back to Sentry. This would indicate a "hole" within this specific trace. The collected spans within the Background Queue would be descendants of a span (for example, a queue service call) within the Backend component, and thus the top-level span of the collective Background Queue spans is referred to as an orphan span.

### Properties

Each span has a unique identifier kept in the `span_id` attribute.

Start time and end time are stored in the `start_timestamp` and `timestamp` attributes, respectively. The format is an RFC 3339 / ISO 8601-compatible string or a numeric value with the number of seconds since the UNIX Epoch.

The name of the operation is stored in the “op” parameter. Examples are `http`, `sql.query`, `redis.command`.

Additional data about the operation can be stored in the `data` and `tags` attributes.

## Trace

> A trace is a collection of spans that have the same `trace_id` attribute. Traces consist of all spans generated by the distributed system during the processing of the input data. Traces are not explicitly started or finished. The start and end of a trace are defined by the transactions and spans participating in the trace.

Author: ^ @dashed

Trace metadata is propagated through http headers that are included in the http request to the external system.

## Transaction

Transaction is an event attribute that provides initial context to the error report. This makes it possible for the error to be linked to a specific part of the system. An example of the transaction name can be “/users/” which is the name for the endpoint that received the request that failed during the processing.

With tracing, the transaction is a parent span to all spans created during the processing of the request in the single server.

Once the request is processed, the transaction is sent together with all child spans to Sentry.

### Event Types

The SDKs already collect a lot of information about the processing of the request. In the case of an error, the SDK attaches the info to the error report to provide rich context of what happened before the error appeared -- information like tags and breadcrumbs.

If the error doesn’t appear during the processing of the request, we discard the context data once processing is complete.

> We want to leverage this context collection to build spans, and once a request is finished, we want to send the data to Sentry.

Reviewer 4: Why are these paragraphs here? The payload below gives an example of a transaction event, but no other event types are covered. If we don't reference this anywhere could we get by without it?

Author: Yeah, I can remove this part. I think this was part of the forum glossary. https://forum.sentry.io/t/sentrys-apm-docs-alpha/7843

Author: @markstory to be clear, we should remove lines 131-133?

Reviewer 4: Unless we need a place to refer people to on the shape of a transaction event, I would remove 127-164. If we do need a reference transaction event, then remove 129-134.

---
title: APM Glossary
sidebar_order: 5
---

### Hierarchy

Each span can be connected to other spans using a parent-child hierarchy. 

For example, a span called `updateJiraTask` would have child spans like `Check permissions`, `Update Task`, `sql.update`, `send.webhooks`.

> Each sub-task would have an id in the attribute `parent_span_id` id of the “updateTask” span.

Reviewer: ```suggestion
Each sub-task span would have an id in the attribute `parent_span_id` id of the “updateTask” span.
```

> Each sub-task would have an id in the attribute `parent_span_id` id of the “updateTask” span.

Reviewer 4: ```suggestion
Each sub-task would have an id in the attribute `parent_span_id` id of the `updateJiraTask` span.
```

Author: @markstory what would be the appropriate slug?

### Properties

Each span has a unique identifier kept in the `span_id` attribute.

Start time and end time are stored in the `start_timestamp` and `timestamp`. End time is captured in the attribute timestamp and not the end_timestamp for now because of compatibility reasons.

The name of the operation is stored in the “op” parameter. Examples are `http`, `sql.query`, `redis.command`.

Additional data about the operation can be stored in the `data` and `tags` attributes.

### Span

The **span** is the primary building block of a distributed trace, representing an individual unit of work done in a distributed system.

It can be the processing of an HTTP request by a server, the querying of a database, or an HTTP call to an external service.

> An example of a span that describes a call to an external service:

Reviewer 3: I think we might be better off having an example that connects to the previous text about `createJiraTask`.

Author: Okay. :) Let me know the best place to get that code example.

    {
          "trace_id": "a0fa8803753e40fd8124b21eeb2986b5",
          "parent_span_id": "9c2a6db8c79068a2",
          "span_id": "8c931f4740435fb8",
          "start_timestamp": 1563890702.134,
          "same_process_as_parent": true,
          "description": "http://httpbin.org/base64/aGVsbG8gd29ybGQK GET",

>           "tags": { "http.status_code": 200, "error": false },

Reviewer 3: It's no longer `"error": false` but rather `"status": "ok"`

>           "tags": { "http.status_code": 200, "error": false },

Reviewer 4: ```suggestion
          "tags": { "http.status_code": 200, "status": "ok" },
```

>           "timestamp": "1563890702.934",

Reviewer 3: Where does this example come from? I do not understand why the timestamp is both a unix timestamp and sent as a string. This seems to be invalid data as per the schema

Author: I got it from the forum docs. Navigate to the bottom Glossary and decompress the toggle: https://forum.sentry.io/t/sentrys-apm-docs-alpha/7843

Reviewer 3: can we still drop the quotes around the number here? This is wrong and is unlikely to be accepted by the server. Fairly certain none of our SDKs send this

Reviewer 3: nvm this is tracked as part of a different convo

Transaction is an event attribute that provides initial context to the error report. This makes it possible for the error to be linked to a specific part of the system. An example of the transaction name can be “/users/” which is the name for the endpoint that received the request that failed during the processing.

With APM, the transaction is a parent span to all spans created during the processing of the request in the single server.

Once the request is processed, the transaction is sent together with all child spans to Sentry.

### Event Types

The SDKs already collect a lot of information about the processing of the request. In the case of an error, the SDK attaches the info to the error report to provide rich context of what happened before the error appeared -- information like tags and breadcrumbs.

If the error doesn’t appear during the processing of the request, we discard the context data once processing is complete.

We want to leverage this context collection to build spans, and once a request is finished, we want to send the data to Sentry.

Example of the transaction event type:

>     {     

Reviewer 3: Can we trim the examples down? A lot of what is contained in this one is irrelevant for APM. I would remove the following keys:

* `modules`
* `extra`
* `server_name`
* `platform`
* `breadcrumbs` (those are not supposed to be sent!)
* `sdk`

            {
              "op": "db",
              "description": "SELECT * from countries where id = ?",
              "start_timestamp": 1562681591,
              "timestamp": 1562681591,
              "parent_span_id": "9c2a6db8c79068a2",
              "trace_id": "a0fa8803753e40fd8124b21eeb2986b5",
              "data": {},
              "tags": {}
            }
          ],
          "breadcrumbs": [
        	  {
        		"category": "httplib",
        	  	"data": {
        		  	"url": "http://httpbin.org/base64/aGVsbG8gd29ybGQK",
        		  	"status_code": 200,
        	  		"reason": "OK",
        	  		"method": "GET"
        		  },
        		  "type": "http",
        		  "timestamp": "2019-06-14T12:01:41Z"
        	  }
          ],
          "sdk": {
        	"version": "0.9.0",
        	"name": "sentry.python",
        	"packages": [{ "version": "0.9.0", "name": "pypi:sentry-sdk" }],
        	"integrations": ["argv"]
          }
        }

## Performance Metric Definitions

As part of APM and transaction data, we’re exposing a few new aggregate metrics to help customers better understand the performance of their applications.

### Duration Percentiles

> The average, 75%, and 95% duration, help guide customers in identifying transactions that are slower than the user’s target SLAs.

Reviewer 4: ```suggestion
The average, 75%, 95%, and 99% duration, help guide customers in identifying transactions that are slower than the user’s target SLAs.
```

Our APM data is sampled, so the percentiles we present only represent the data we have. Because of sampling, data ranges, filters, or a low traffic transaction, we will often have a case where data is directionally correct, but not accurate. For example, the average of one number is that number, but that’s not usually what people expect to see.

> This inability to be usefully accurate will happen more often in some columns than others. We can calculate an average with less data than a 95th percentile, but it’ll also vary by row. An index page will always get more traffic than /settings/country/belgium/tax.

Reviewer 4: ```suggestion
This inability to be usefully accurate will happen more often in some columns than others. We can calculate an average with less data than a 95th percentile, but it’ll also vary by row. An index page will always get more traffic than `/settings/country/belgium/tax`.
```

Should URLs be formatted as monospace?

To signal this lack of accuracy to the user, we’ll show a **warning** to users:

- If there are fewer than 5 data-points, we’ll show “n/a” in the average column.
- If there are fewer than 20 data-points, we’ll show “n/a” in the 75th percentile
- If there are fewer than 100 data-points, we’ll show “n/a” in the 95th percentile

> The UI will highlight any warnings visually and let the user know that they should increase their sampling rate or make a less precise query.

Reviewer 4: We never built any of these warnings :blush:
